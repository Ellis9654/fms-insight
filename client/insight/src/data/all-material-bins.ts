/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import * as api from "./api";
import * as im from "immutable";

export type MaterialList = ReadonlyArray<Readonly<api.IInProcessMaterial>>;

export type AllMaterialBins = im.Map<string, MaterialList>;

export function selectAllMaterialIntoBins(curSt: Readonly<api.ICurrentStatus>): AllMaterialBins {
  const palLoc = im
    .Map(curSt.pallets)
    .map(st => " (" + st.currentPalletLocation.group + " #" + st.currentPalletLocation.num.toString() + ")");

  return im
    .Seq(curSt.material)
    .map(mat => {
      switch (mat.location.type) {
        case api.LocType.InQueue:
          return { region: mat.location.currentQueue || "Queue", mat };
        case api.LocType.OnPallet:
          let region = "Pallet";
          if (mat.location.pallet) {
            region = "Pallet " + mat.location.pallet.toString() + palLoc.get(mat.location.pallet) || "";
          }
          return { region, mat };
        case api.LocType.Free:
        default:
          switch (mat.action.type) {
            case api.ActionType.Loading:
              return { region: "Raw Material", mat };
            default:
              return { region: "Free Material", mat };
          }
      }
    })
    .groupBy(x => x.region)
    .map(group =>
      group
        .map(x => x.mat)
        .sortBy(mat => {
          switch (mat.location.type) {
            case api.LocType.OnPallet:
              return mat.location.face;
            case api.LocType.InQueue:
            case api.LocType.Free:
            default:
              return mat.location.queuePosition;
          }
        })
        .valueSeq()
        .toArray()
    )
    .toMap();
}
