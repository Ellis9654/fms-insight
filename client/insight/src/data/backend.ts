/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import * as api from "./api";
import * as im from "immutable";
import { User } from "oidc-client";

export interface JobAPI {
  history(startUTC: Date, endUTC: Date): Promise<Readonly<api.IHistoricData>>;
  currentStatus(): Promise<Readonly<api.ICurrentStatus>>;
  mostRecentUnfilledWorkordersForPart(part: string): Promise<ReadonlyArray<Readonly<api.IPartWorkorder>>>;

  removeMaterialFromAllQueues(materialId: number): Promise<void>;
  setMaterialInQueue(materialId: number, queue: api.QueuePosition): Promise<void>;
  addUnprocessedMaterialToQueue(
    jobUnique: string,
    lastCompletedProcess: number,
    queue: string,
    pos: number,
    serial: string
  ): Promise<void>;
}

export interface ServerAPI {
  fMSInformation(): Promise<Readonly<api.IFMSInfo>>;
}

export interface LogAPI {
  get(startUTC: Date, endUTC: Date): Promise<ReadonlyArray<Readonly<api.ILogEntry>>>;
  recent(lastSeenCounter: number): Promise<ReadonlyArray<Readonly<api.ILogEntry>>>;
  logForMaterial(materialID: number): Promise<ReadonlyArray<Readonly<api.ILogEntry>>>;
  logForSerial(serial: string): Promise<ReadonlyArray<Readonly<api.ILogEntry>>>;
  getWorkorders(ids: string[]): Promise<ReadonlyArray<Readonly<api.IWorkorderSummary>>>;

  setInspectionDecision(inspType: string, mat: api.LogMaterial, inspect: boolean): Promise<Readonly<api.ILogEntry>>;
  recordInspectionCompleted(insp: api.NewInspectionCompleted): Promise<Readonly<api.ILogEntry>>;
  recordWashCompleted(insp: api.NewWash): Promise<Readonly<api.ILogEntry>>;
  setWorkorder(workorder: string, mat: api.LogMaterial): Promise<Readonly<api.ILogEntry>>;
  setSerial(serial: string, mat: api.LogMaterial): Promise<Readonly<api.ILogEntry>>;
}

export const BackendHost = process.env.NODE_ENV === "production" ? undefined : "localhost:5000";
const BackendUrl = BackendHost ? "http://" + BackendHost : undefined;
// tslint:disable-next-line:no-any
export const DemoMode = (window as any).FMS_INSIGHT_DEMO_MODE || false;

export let ServerBackend: ServerAPI = new api.ServerClient(BackendUrl);
export let JobsBackend: JobAPI = new api.JobsClient(BackendUrl);
export let LogBackend: LogAPI = new api.LogClient(BackendUrl);
let otherLogServers: ReadonlyArray<string> = [];
export let OtherLogBackends: ReadonlyArray<LogAPI> = [];

export function setOtherLogBackends(servers: ReadonlyArray<string>) {
  otherLogServers = servers;
  OtherLogBackends = servers.map(s => new api.LogClient(s));
}

export function setUserToken(u: User) {
  const token = u.access_token || u.id_token;
  function fetch(url: RequestInfo, init?: RequestInit) {
    return window.fetch(
      url,
      init
        ? { ...init, headers: { ...init.headers, Authorization: "Bearer " + token } }
        : { headers: { Authorization: "Bearer " + token } }
    );
  }
  ServerBackend = new api.ServerClient(BackendUrl, { fetch });
  JobsBackend = new api.JobsClient(BackendUrl, { fetch });
  LogBackend = new api.LogClient(BackendUrl, { fetch });
  OtherLogBackends = otherLogServers.map(s => new api.LogClient(s, { fetch }));
}

export interface MockData {
  readonly curSt: Readonly<api.ICurrentStatus>;
  readonly jobs: Readonly<api.IHistoricData>;
  readonly workorders: Map<string, ReadonlyArray<Readonly<api.IPartWorkorder>>>;
  readonly events: Promise<Readonly<api.ILogEntry>[]>;
}

function initMockBackend(data: Promise<MockData>) {
  ServerBackend = {
    fMSInformation() {
      return Promise.resolve({
        name: "mock",
        version: "1.0.0",
        requireScanAtWash: false,
        requireWorkorderBeforeAllowWashComplete: false,
        additionalLogServers: []
      });
    }
  };

  JobsBackend = {
    history(startUTC: Date, endUTC: Date): Promise<Readonly<api.IHistoricData>> {
      return data.then(d => d.jobs);
    },
    currentStatus(): Promise<Readonly<api.ICurrentStatus>> {
      return data.then(d => d.curSt);
    },
    mostRecentUnfilledWorkordersForPart(part: string): Promise<ReadonlyArray<Readonly<api.IPartWorkorder>>> {
      return data.then(d => d.workorders.get(part) || []);
    },

    removeMaterialFromAllQueues(materialId: number): Promise<void> {
      // do nothing
      return Promise.resolve();
    },
    setMaterialInQueue(materialId: number, queue: api.QueuePosition): Promise<void> {
      // do nothing
      return Promise.resolve();
    },
    addUnprocessedMaterialToQueue(
      jobUnique: string,
      lastCompletedProcess: number,
      queue: string,
      pos: number,
      serial: string
    ): Promise<void> {
      // do nothing
      return Promise.resolve();
    }
  };

  const serialsToMatId = data.then(d =>
    d.events.then(evts =>
      im.Map(
        im
          .Seq(evts)
          .filter(e => e.type === api.LogType.PartMark)
          .flatMap(e => e.material.map(m => [e.result, m.id] as [string, number]))
      )
    )
  );

  LogBackend = {
    get(startUTC: Date, endUTC: Date): Promise<ReadonlyArray<Readonly<api.ILogEntry>>> {
      return data.then(d =>
        d.events.then(evts =>
          im
            .Seq(evts)
            .filter(e => e.endUTC >= startUTC && e.endUTC <= endUTC)
            .toArray()
        )
      );
    },
    recent(lastSeenCounter: number): Promise<ReadonlyArray<Readonly<api.ILogEntry>>> {
      // no recent events, everything is static
      return Promise.resolve([]);
    },
    logForMaterial(materialID: number): Promise<ReadonlyArray<Readonly<api.ILogEntry>>> {
      return data.then(d =>
        d.events.then(evts =>
          im
            .Seq(evts)
            .filter(e => im.Seq(e.material).some(m => m.id === materialID))
            .toArray()
        )
      );
    },
    logForSerial(serial: string): Promise<ReadonlyArray<Readonly<api.ILogEntry>>> {
      return serialsToMatId.then(s => {
        var mId = s.get(serial);
        if (mId) {
          return this.logForMaterial(mId);
        } else {
          return Promise.resolve([]);
        }
      });
    },
    getWorkorders(ids: string[]): Promise<ReadonlyArray<Readonly<api.IWorkorderSummary>>> {
      // no workorder summaries
      return Promise.resolve([]);
    },

    setInspectionDecision(inspType: string, mat: api.LogMaterial, inspect: boolean): Promise<Readonly<api.ILogEntry>> {
      const evt = {
        counter: 0,
        material: [mat],
        pal: "",
        type: api.LogType.Inspection,
        startofcycle: false,
        endUTC: new Date(),
        loc: "Inspection",
        locnum: 1,
        result: inspect.toString(),
        program: "",
        elapsed: "00:00:00",
        active: "00:00:00",
        details: {
          InspectionType: inspType
        }
      };
      return data.then(d =>
        d.events.then(evts => {
          evts.push(evt);
          return evt;
        })
      );
    },
    recordInspectionCompleted(insp: api.NewInspectionCompleted): Promise<Readonly<api.ILogEntry>> {
      const evt: api.ILogEntry = {
        counter: 0,
        material: [insp.material],
        pal: "",
        type: api.LogType.InspectionResult,
        startofcycle: false,
        endUTC: new Date(),
        loc: "InspectionComplete",
        locnum: insp.inspectionLocationNum,
        result: insp.success.toString(),
        program: insp.inspectionType,
        elapsed: insp.elapsed,
        active: insp.active,
        details: insp.extraData
      };
      return data.then(d =>
        d.events.then(evts => {
          evts.push(evt);
          return evt;
        })
      );
    },
    recordWashCompleted(wash: api.NewWash): Promise<Readonly<api.ILogEntry>> {
      const evt: api.ILogEntry = {
        counter: 0,
        material: [wash.material],
        pal: "",
        type: api.LogType.Wash,
        startofcycle: false,
        endUTC: new Date(),
        loc: "Wash",
        locnum: wash.washLocationNum,
        result: "",
        program: "",
        elapsed: wash.elapsed,
        active: wash.active,
        details: wash.extraData
      };
      return data.then(d =>
        d.events.then(evts => {
          evts.push(evt);
          return evt;
        })
      );
    },
    setWorkorder(workorder: string, mat: api.LogMaterial): Promise<Readonly<api.ILogEntry>> {
      const evt: api.ILogEntry = {
        counter: 0,
        material: [mat],
        pal: "",
        type: api.LogType.OrderAssignment,
        startofcycle: false,
        endUTC: new Date(),
        loc: "OrderAssignment",
        locnum: 1,
        result: workorder,
        program: "",
        elapsed: "00:00:00",
        active: "00:00:00"
      };
      return data.then(d =>
        d.events.then(evts => {
          evts.push(evt);
          return evt;
        })
      );
    },
    setSerial(serial: string, mat: api.LogMaterial): Promise<Readonly<api.ILogEntry>> {
      const evt: api.ILogEntry = {
        counter: 0,
        material: [mat],
        pal: "",
        type: api.LogType.PartMark,
        startofcycle: false,
        endUTC: new Date(),
        loc: "Mark",
        locnum: 1,
        result: serial,
        program: "",
        elapsed: "00:00:00",
        active: "00:00:00"
      };
      return data.then(d =>
        d.events.then(evts => {
          evts.push(evt);
          return evt;
        })
      );
    }
  };
}

export function registerMockBackend() {
  const mockDataPromise = new Promise<MockData>(function(resolve: (d: MockData) => void) {
    // tslint:disable-next-line:no-any
    (window as any).FMS_INSIGHT_RESOLVE_MOCK_DATA = resolve;
  });
  initMockBackend(mockDataPromise);
}
