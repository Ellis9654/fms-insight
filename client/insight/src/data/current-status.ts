/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
import * as api from './api';
import { ConsumingPledge, PledgeStatus } from './pledge';

export interface State {
    readonly loading: boolean;
    readonly loading_error?: Error;
    readonly current_status: Readonly<api.ICurrentStatus>; // TODO: DeepReadonly
}

const initial: State = {
    loading: false,
    current_status: {
        jobs: {},
        pallets: {},
        material: [],
        alarms: [],
        queues: {}
    }
};

export enum ActionType {
    NewCurrentStatus = 'CurStatus_NewCurrentStatus',
    Other = 'Other',
}

// TODO: Pledge and DeepReadonly in typescript 2.8
export type Action =
  | {type: ActionType.NewCurrentStatus, pledge: ConsumingPledge<Readonly<api.ICurrentStatus>> }
  | {type: ActionType.Other}
  ;

export function loadCurrentStatus() {
    var client = new api.JobsClient();
    return {
        type: ActionType.NewCurrentStatus,
        pledge: client.currentStatus()
    };
}

export function reducer(s: State, a: Action): State {
    if (s === undefined) { return initial; }
    switch (a.type) {
        case ActionType.NewCurrentStatus:
            switch (a.pledge.status) {
                case PledgeStatus.Starting:
                    return {...s, loading: true, loading_error: undefined};
                case PledgeStatus.Completed:
                    return {...s,
                        loading: false,
                        current_status: a.pledge.result,
                    };
                case PledgeStatus.Error:
                    return {...s, loading_error: a.pledge.error};

                default: return s;
            }
        default: return s;
    }
}