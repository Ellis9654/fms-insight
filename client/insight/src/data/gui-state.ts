/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

export enum PlannedOrActual {
  Planned = 'Planned',
  Actual = 'Actual',
  PlannedMinusActual = 'PlannedOrActual',
}

export enum AddMatToQueueDialogState {
  DialogClosed,
  DialogOpenChooseSerialOrJob,
  DialogOpenToAddMaterial,
}

export enum ActionType {
  SetSelectedStationCyclePart = 'Gui_SetSelectedStationCyclePart',
  SetSelectedPalletCycle = 'Gui_SetSelectedPalletCycle',
  SetStationOeeHeatmapType = 'Gui_SetStationOeeHeatmapType',
  SetCompletedCountHeatmapType = 'Gui_SetCompletedCountHeatmapType',
  SetWorkorderDialogOpen = 'Gui_SetWorkorderDialog',
  SetInspTypeDialogOpen = 'Gui_SetInspTypeDialog',
  SetSerialDialogOpen = 'Gui_SetSerialDialogOpen',
  SetAddMatToQueueDialog = 'Gui_SetAddMatToQueueDialog',
  SetScanQrCodeDialog = 'Gui_ScanQrCodeDialog',
}

export type Action =
  | { type: ActionType.SetSelectedStationCyclePart, part: string }
  | { type: ActionType.SetSelectedPalletCycle, pallet: string }
  | { type: ActionType.SetStationOeeHeatmapType, ty: PlannedOrActual }
  | { type: ActionType.SetCompletedCountHeatmapType, ty: PlannedOrActual }
  | { type: ActionType.SetWorkorderDialogOpen, open: boolean }
  | { type: ActionType.SetInspTypeDialogOpen, open: boolean }
  | { type: ActionType.SetSerialDialogOpen, open: boolean }
  | { type: ActionType.SetAddMatToQueueDialog, queue?: string, st: AddMatToQueueDialogState }
  | { type: ActionType.SetScanQrCodeDialog, open: boolean }
  ;

export interface State {
  readonly station_cycle_selected_part?: string;
  readonly pallet_cycle_selected?: string;
  readonly station_oee_heatmap_type: PlannedOrActual;
  readonly completed_count_heatmap_type: PlannedOrActual;
  readonly workorder_dialog_open: boolean;
  readonly insptype_dialog_open: boolean;
  readonly serial_dialog_open: boolean;
  readonly scan_qr_dialog_open: boolean;
  readonly add_mat_to_queue_st: AddMatToQueueDialogState;
  readonly add_mat_to_queue?: string;
}

export const initial: State = {
  station_oee_heatmap_type: PlannedOrActual.Actual,
  completed_count_heatmap_type: PlannedOrActual.Actual,
  workorder_dialog_open: false,
  insptype_dialog_open: false,
  serial_dialog_open: false,
  scan_qr_dialog_open: false,
  add_mat_to_queue_st: AddMatToQueueDialogState.DialogClosed,
};

export function reducer(s: State, a: Action): State {
  if (s === undefined) { return initial; }
  switch (a.type) {
    case ActionType.SetSelectedStationCyclePart:
      return {...s, station_cycle_selected_part: a.part };
    case ActionType.SetSelectedPalletCycle:
      return {...s, pallet_cycle_selected: a.pallet };
    case ActionType.SetStationOeeHeatmapType:
      return {...s, station_oee_heatmap_type: a.ty };
    case ActionType.SetCompletedCountHeatmapType:
      return {...s, completed_count_heatmap_type: a.ty };
    case ActionType.SetWorkorderDialogOpen:
      return {...s, workorder_dialog_open: a.open };
    case ActionType.SetInspTypeDialogOpen:
      return {...s, insptype_dialog_open: a.open };
    case ActionType.SetSerialDialogOpen:
      return {...s, serial_dialog_open: a.open };
    case ActionType.SetScanQrCodeDialog:
      return {...s, scan_qr_dialog_open: a.open };
    case ActionType.SetAddMatToQueueDialog:
      return {...s, add_mat_to_queue_st: a.st, add_mat_to_queue: a.queue };
    default:
      return s;
  }
}