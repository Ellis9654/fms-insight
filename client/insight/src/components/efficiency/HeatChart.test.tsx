/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import * as React from 'react';
import { shallow } from 'enzyme';
import WorkIcon from 'material-ui-icons/Work';

import { HeatChart, HeatChartProps, SelectableHeatChart, SelectableHeatChartProps } from './HeatChart';

import * as gui from '../../data/gui-state';

const points = [
  {x: new Date(2018, 2, 1), y: "aaa", color: 1},
  {x: new Date(2018, 2, 2), y: "aaa", color: 2},
  {x: new Date(2018, 2, 3), y: "aaa", color: 3},
  {x: new Date(2018, 2, 1), y: "bbb", color: 4},
  {x: new Date(2018, 2, 2), y: "bbb", color: 5},
  {x: new Date(2018, 2, 3), y: "bbb", color: 6},
];

it('displays a cycle chart', () => {
  const props: HeatChartProps = {
    points: points,
    color_label: "colorlabel",
    row_count: 10,
  };

  const val = shallow(<HeatChart {...props}/>);
  expect(val).toMatchSnapshot('heat chart');
});

it('displays a selectable heat chart with actual points', () => {
  const props: SelectableHeatChartProps = {
    points: points,
    color_label: "colorlabel",
    card_label: "cardlabel",
    icon: <WorkIcon/>,
    planned_or_actual: gui.PlannedOrActual.Actual,
    setType: jest.fn()
  };

  const val = shallow(<SelectableHeatChart {...props}/>);
  expect(val).toMatchSnapshot('selectable heat chart with actual data');
});

it('displays a selectable heat chart with planned points', () => {
  const props: SelectableHeatChartProps = {
    points: points,
    color_label: "colorlabel",
    card_label: "cardlabel",
    icon: <WorkIcon/>,
    planned_or_actual: gui.PlannedOrActual.Planned,
    setType: jest.fn()
  };

  const val = shallow(<SelectableHeatChart {...props}/>);
  expect(val).toMatchSnapshot('selectable heat chart with planned data');
});