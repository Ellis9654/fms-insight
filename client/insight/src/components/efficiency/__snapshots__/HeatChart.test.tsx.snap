// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`heat chart 1`] = `
<FlexibleXYPlot
  colorRange={
    Array [
      "#E8F5E9",
      "#1B5E20",
    ]
  }
  height={750}
  margin={
    Object {
      "bottom": 60,
      "left": 100,
    }
  }
  xType="ordinal"
  yType="ordinal"
>
  <XAxis
    attr="x"
    attrAxis="y"
    orientation="bottom"
    tickFormat={[Function]}
    tickLabelAngle={-45}
  />
  <YAxis
    attr="y"
    attrAxis="x"
    orientation="left"
  />
  <HeatmapSeries
    className=""
    data={
      Array [
        Object {
          "color": 1,
          "label": "1",
          "x": 2018-03-01T06:00:00.000Z,
          "y": "aaa",
        },
        Object {
          "color": 2,
          "label": "2",
          "x": 2018-03-02T06:00:00.000Z,
          "y": "aaa",
        },
        Object {
          "color": 3,
          "label": "3",
          "x": 2018-03-03T06:00:00.000Z,
          "y": "aaa",
        },
        Object {
          "color": 4,
          "label": "4",
          "x": 2018-03-01T06:00:00.000Z,
          "y": "bbb",
        },
        Object {
          "color": 5,
          "label": "5",
          "x": 2018-03-02T06:00:00.000Z,
          "y": "bbb",
        },
        Object {
          "color": 6,
          "label": "6",
          "x": 2018-03-03T06:00:00.000Z,
          "y": "bbb",
        },
      ]
    }
    onValueMouseOut={[Function]}
    onValueMouseOver={[Function]}
    stack={false}
    style={Object {}}
  />
</FlexibleXYPlot>
`;

exports[`selectable heat chart with actual data 1`] = `
<Card
  raised={true}
>
  <WithStyles(CardHeader)
    title={
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexWrap": "wrap",
          }
        }
      >
        <pure(Work) />
        <div
          style={
            Object {
              "marginLeft": "10px",
              "marginRight": "3em",
            }
          }
        >
          cardlabel
        </div>
        <div
          style={
            Object {
              "flexGrow": 1,
            }
          }
        />
        <WithStyles(Select)
          autoWidth={true}
          displayEmpty={true}
          onChange={[Function]}
          value="Actual"
        >
          <WithStyles(MenuItem)
            value="Actual"
          >
            Actual
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="Planned"
          >
            Planned
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="PlannedOrActual"
          >
            Planned minus Actual
          </WithStyles(MenuItem)>
        </WithStyles(Select)>
      </div>
    }
  />
  <WithStyles(CardContent)>
    <HeatChart
      label_title="colorlabel"
      points={
        Array [
          Object {
            "color": 1,
            "label": "1",
            "x": 2018-03-01T06:00:00.000Z,
            "y": "aaa",
          },
          Object {
            "color": 2,
            "label": "2",
            "x": 2018-03-02T06:00:00.000Z,
            "y": "aaa",
          },
          Object {
            "color": 3,
            "label": "3",
            "x": 2018-03-03T06:00:00.000Z,
            "y": "aaa",
          },
          Object {
            "color": 4,
            "label": "4",
            "x": 2018-03-01T06:00:00.000Z,
            "y": "bbb",
          },
          Object {
            "color": 5,
            "label": "5",
            "x": 2018-03-02T06:00:00.000Z,
            "y": "bbb",
          },
          Object {
            "color": 6,
            "label": "6",
            "x": 2018-03-03T06:00:00.000Z,
            "y": "bbb",
          },
        ]
      }
      row_count={2}
    />
  </WithStyles(CardContent)>
</Card>
`;

exports[`selectable heat chart with planned data 1`] = `
<Card
  raised={true}
>
  <WithStyles(CardHeader)
    title={
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexWrap": "wrap",
          }
        }
      >
        <pure(Work) />
        <div
          style={
            Object {
              "marginLeft": "10px",
              "marginRight": "3em",
            }
          }
        >
          cardlabel
        </div>
        <div
          style={
            Object {
              "flexGrow": 1,
            }
          }
        />
        <WithStyles(Select)
          autoWidth={true}
          displayEmpty={true}
          onChange={[Function]}
          value="Planned"
        >
          <WithStyles(MenuItem)
            value="Actual"
          >
            Actual
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="Planned"
          >
            Planned
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="PlannedOrActual"
          >
            Planned minus Actual
          </WithStyles(MenuItem)>
        </WithStyles(Select)>
      </div>
    }
  />
  <WithStyles(CardContent)>
    <HeatChart
      label_title="colorlabel"
      points={
        Array [
          Object {
            "color": 1,
            "label": "1",
            "x": 2018-03-01T06:00:00.000Z,
            "y": "aaa",
          },
          Object {
            "color": 2,
            "label": "2",
            "x": 2018-03-02T06:00:00.000Z,
            "y": "aaa",
          },
          Object {
            "color": 3,
            "label": "3",
            "x": 2018-03-03T06:00:00.000Z,
            "y": "aaa",
          },
          Object {
            "color": 4,
            "label": "4",
            "x": 2018-03-01T06:00:00.000Z,
            "y": "bbb",
          },
          Object {
            "color": 5,
            "label": "5",
            "x": 2018-03-02T06:00:00.000Z,
            "y": "bbb",
          },
          Object {
            "color": 6,
            "label": "6",
            "x": 2018-03-03T06:00:00.000Z,
            "y": "bbb",
          },
        ]
      }
      row_count={2}
    />
  </WithStyles(CardContent)>
</Card>
`;
